Missing form data

While testing I ran in to an issue a few times while trying to insert a note. The form wasn't being created properly so I was only presented with an empty dialog box. After pressing okay on the box I was given this error in the console.
```
plugin_joplin.plugin.templates.js:273 Error in parsing template. TypeError: Cannot read property 'variables' of undefined
    at t.Parser.<anonymous> (plugin_joplin.plugin.templates.js:273)
    at Generator.next (<anonymous>)
    at i (plugin_joplin.plugin.templates.js:273)
```
I can't reproduce this consistently, but it did happen to me about 4 times during testing. I'll try to get you more information that will help you troubleshoot this. I suspect this might be a bug in the core app, so it could be awhile to find it.

Here is the template that caused it (Not consistently)
```
---
title: text
date: text
hours: number
something: number

---

Date: {{date}}
Title: {{title}}
Hours: {{hours}}
```

--- a/src/index.ts
+++ b/src/index.ts
@@ -7,6 +7,7 @@ import { setDefaultTemplatesView } from "./views/defaultTemplates";
 import { TemplateAction, performAction } from "./actions";
 import { loadLegacyTemplates } from "./legacyTemplates";
 import * as open from "open";
+import { Logger } from "./logger";
 
 const DOCUMENTATION_URL = "https://github.com/joplin/plugin-templates#readme";
 
@@ -35,12 +36,15 @@ joplin.plugins.register({
         const userLocale = await joplin.settings.globalValue("locale");
         const userDateFormat = await joplin.settings.globalValue("dateFormat");
         const userTimeFormat = await joplin.settings.globalValue("timeFormat");
+        const profileDir = await joplin.settings.globalValue("profileDir");
+
         const dateAndTimeUtils = new DateAndTimeUtils(userLocale, userDateFormat, userTimeFormat);
-        const parser = new Parser(dateAndTimeUtils, dialogViewHandle);
+        const logger = new Logger(profileDir);
+        const parser = new Parser(dateAndTimeUtils, dialogViewHandle, logger);
 
 
         // Asynchronously load legacy templates
-        loadLegacyTemplates(dateAndTimeUtils);
+        loadLegacyTemplates(dateAndTimeUtils, profileDir);
 
 
         // Utility Functions

--- a/src/legacyTemplates.ts
+++ b/src/legacyTemplates.ts
@@ -22,13 +22,12 @@ const getTemplatesTag = async (): Promise<string> => {
     return (await getAnyTagWithTitle("template")).id;
 }
 
-export const loadLegacyTemplates = async (dateAndTimeUtils: DateAndTimeUtils): Promise<void> => {
+export const loadLegacyTemplates = async (dateAndTimeUtils: DateAndTimeUtils, profileDir: string): Promise<void> => {
     const fs = joplin.require("fs-extra");
 
     let folderId = null;
     let templatesTagId = null;
 
-    const profileDir = await joplin.settings.globalValue("profileDir");
     const templatesDir = `${profileDir}/templates`;
 
     if (await fs.pathExists(templatesDir)) {

--- a/src/logger.ts
+++ b/src/logger.ts
@@ -1,0 +1,15 @@
+import joplin from "api";
+
+const fs = joplin.require("fs-extra");
+
+export class Logger {
+    private logsFile: string;
+
+    constructor(profileDir: string) {
+        this.logsFile = `${profileDir}/templates-logs.txt`;
+    }
+
+    public async log(message: string): Promise<void> {
+        await fs.appendFile(this.logsFile, `[${new Date().toISOString()}]\n${message}\n\n\n`);
+    }
+}

--- a/src/parser.ts
+++ b/src/parser.ts
@@ -1,5 +1,6 @@
 import joplin from "api";
 import * as Handlebars from "handlebars/dist/handlebars";
+import { Logger } from "./logger";
 import { DateAndTimeUtils } from "./utils/dateAndTime";
 import { Note } from "./utils/templates";
 import { setTemplateVariablesView } from "./views/templateVariables";
@@ -13,10 +14,12 @@ const frontmatter = require("front-matter");
 export class Parser {
     private utils: DateAndTimeUtils;
     private dialog: string;
+    private logger: Logger;
 
-    constructor(dateAndTimeUtils: DateAndTimeUtils, dialogViewHandle: string) {
+    constructor(dateAndTimeUtils: DateAndTimeUtils, dialogViewHandle: string, logger: Logger) {
         this.utils = dateAndTimeUtils;
         this.dialog = dialogViewHandle;
+        this.logger = logger;
     }
 
     private getDefaultContext() {
@@ -68,7 +71,24 @@ export class Parser {
             return null;
         }
 
-        const userResponse = dialogResponse.formData.variables;
+        let userResponse;
+
+        // There's a try catch block here because a user experienced an error
+        // due to the following line. I've added a try catch block to log the
+        // necessary info to find the root cause of the error in case it happens again.
+        // Reference -> https://github.com/joplin/plugin-templates/issues/6
+        try {
+            userResponse = dialogResponse.formData.variables;
+        } catch (err) {
+            console.error("Template variables form was not able to load properly.", err);
+            console.error("DEBUG INFO", variables, dialogResponse, title);
+
+            const message = (`Template variables form was not able to load properly.\n${err}\nDEBUG INFO\nvariables: ${JSON.stringify(variables)}\ndialogResponse: ${JSON.stringify(dialogResponse)}\ntitle: ${JSON.stringify(title)}`);
+            this.logger.log(message);
+
+            throw new Error(err);
+        }
+
         return this.mapUserResponseToVariables(variables, userResponse);
     }
 

